Enumerable module

sz16900 opened the pull request:
# Completed [Project # 9 Bubble Sort](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks?ref=lnav) in partial requirements of the [Microverse Curriculum](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/57421588/).

---

### Contributors:
- Seth Zea

--- 

### Targets Achieved 

##### General Requirements

- [x] Add a descriptive README file to your repository.
- [x] Use meaningful names for your Pull Requests and add a short summary of its content.
- [x] Use descriptive commit messages.
- [x] Do not commit files that are not related to the project. Instead, add them to .gitignore file.
- [x] Do not commit inline comments. 
- [x] Use stickler for checks.

##### Other Requirements
- [x] Keep your code clean.
- [x] Do not commit debugging instructions

##### Project Specific
- [x] Remember that you should use the Git Flow for the first time. Pay attention to the “Remember to” and “Code review” sections.
- [x] Take a look at Ruby docs about Enumerable class https://ruby-doc.org/core-2.6.5/Enumerable.html

# Project Status :
<img src="https://media.giphy.com/media/8JW82ndaYfmNoYAekM/giphy.gif" width="100%">

sz16900 committed:
:sparkles: my_each

add the my_each feature with testing

sz16900 committed:
:sparkles: my_each_with_index

add my_each with index

sz16900 committed:
:rotating_light: add linter files

rubocop + stickler

sz16900 committed:
:sparkles: my_select

add my_select, refactored code, add more tests, add return enumerables

sz16900 committed:
:sparkles: my_all?

add my_all? feature. Probably needs a refactor. There is some repetitive code.

sz16900 committed:
:sparkles: add-my_any?

add my_any? feature to Enumerator module. Need to take care of Regex ===

sz16900 committed:
:bug: fixing bug

some refactor. deleting redundant code

sz16900 committed:
:sparkles: add my_count

add my_count feature, returns specific according to documentation

sz16900 committed:
:sparkles: add my_map

add my map to the enumerator module with testing

sz16900 committed:
:sparkles: add my_inject

add my_inject. feature is too high complexity. needs refactoring. good testing

sz16900 committed:
add content + testing

sz16900 committed:
:recycle: refactor my_inject, my_all?, my_any?

take some code out which was redundant and was adding cyclomatic complexity to my code

sz16900 committed:
:recycle: refactor my_each_with_index

add customizable index

sz16900 committed:
:recycle: refactor my_inject

fewer lines of code by refactor

sz16900 committed:
:rotating_light: remove linter warning

remove linter warnings by making code smaller, cleaner, more readable

sz16900 committed:
Update README.md

sz16900 committed:
:fire: remove testing file

remove testing file

sz16900 committed:
Merge branch 'enumerable_module' of https://github.com/sz16900/enumerables into enumerable_module

sz16900 committed:
Update README.md

sz16900 committed:
Update README.md

whiz25 reviewed (changes requested):
## Nice progress :+1: @sz16900 
You are almost there. We need to fix a few things before we can merge this one.
Check my comments on how to fix them.

### Martin Afani

whiz25 commented on enumerables.rb:
```diff
@@ -0,0 +1,145 @@
+      return false unless yield x
+    end
+    true
+  end
```
- [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true when none of the collection members are false or nil
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

Sample cases:
```ruby
let(:true_array) { [1, true, 'hi', []] }
let(:false_array) { [1, false, 'hi', []] }

expect(true_array.my_all?).to be true_array.all?
expect(false_array.my_all?).to be false_array.all?
```

whiz25 commented on enumerables.rb:
```diff
@@ -0,0 +1,145 @@
+      return true if yield x
+    end
+    false
+  end
```
- [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

Sample testcase:
```ruby
let(:true_array) { [nil, false, true, []] }
let(:false_array) { [nil, false, nil, false] }

expect(true_array.my_any?).to be true_array.any?
expect(false_array.my_any?).to be false_array.any?
```
Should return `true` if `my_any?` works well

whiz25 commented on enumerables.rb:
```diff
@@ -0,0 +1,145 @@
+      return false if yield x
+    end
+    true
+  end
```
- [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true only if none of the collection members is true
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

Sample test case:
```ruby
let(:true_array) { [nil, false, true, []] }
let(:false_array) { [nil, false, nil, false] }

expect(false_array.my_none?).to be true
expect(true_array.my_none?).to be false
```
Should return `true` and `false`

whiz25 commented on enumerables.rb:
```diff
@@ -0,0 +1,145 @@
+    end
+    repetitions += 1 while repetitions < size
+    repetitions
+  end
```
- [ ] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns the number of items in enum through enumeration
- [x] counts the number of items in enum that are equal to item if an argument is given
- [x] counts the number of elements yielding a true value if a block is given

Sample test case:
```ruby
let(:array) { Array.new(100) { rand(0...9) } }
expect(array.my_count).to eq array.count
```
Should return `true`

whiz25 commented on enumerables.rb:
```diff
@@ -0,0 +1,145 @@
+             end
+    end
+    arr
+  end
```
- [ ] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an Enumerator if no block is given
Sample test case:
```ruby
let(:array) { Array.new(100) { rand(0...9) } }
expect(array.my_map).to be_an(Enumerator)
```
Should return `true`

sz16900 committed:
:recycle: change my_all? + my_map

there were some bugs that were taken care of

sz16900 committed:
:bug: fix my_all?

 when no block or argument is given returns true if at least one of the collection is not false or nil

sz16900 committed:
:bug: fix my_any?

when no block or argument is given returns true only if none of the collection members is true

sz16900 committed:
:bug: fix my_count

 returns the number of items in enum through enumeration

sz16900 committed:
:rotating_light: lint my_any?

removing linter warnings, making code cleaner

sz16900 committed:
:fire: remove testing file

remove testing file

sz16900 committed:
Merge branch 'enumerable_module' of https://github.com/sz16900/enumerables into enumerable_module

Forison reviewed (approved):
## Status: Approved💚 
# 🙌 Zea,
Good job on implementing the changes suggested earlier, go ahead and merge.
## Proceed to the next task👌.
### Happy coding💻
### Addo Forison
